### PR TITLE
[multistage]Refactor Mailbox assignment logic to planner

### PR DESCRIPTION
**Pull request by xiangfu0**
1. Move mailbox assignment logic from runtime to planner.
2. Make StagePlan only carries its own StageMetadata.
3. WorkerMetadata has a map from the stageId to the list of mailboxes its connecting to(no need to distinguish sender/receive).

### CODE REVIEW

File: pinot-query-planner/src/main/java/org/apache/pinot/query/routing/StageMetadata.java [hunk 1]
```diff
@@ -29,10 +29,13 @@
  */
 public class StageMetadata {
   private final List<WorkerMetadata> _workerMetadataList;
+  private final Map<Integer, List<MailboxInfo>> _mailBoxInfosMap;
   private final Map<String, String> _customProperties;
 
-  public StageMetadata(List<WorkerMetadata> workerMetadataList, Map<String, String> customProperties) {
+  public StageMetadata(List<WorkerMetadata> workerMetadataList, Map<Integer, List<MailboxInfo>> mailBoxInfosMap,
```
Comment (walterddr) [inline pinot-query-planner/src/main/java/org/apache/pinot/query/routing/StageMetadata.java:35]: MailboxInfo should be on WorkerMetadata level

File: pinot-query-planner/src/main/java/org/apache/pinot/query/planner/physical/MailboxAssignmentVisitor.java [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+import org.apache.pinot.query.routing.QueryServerInstance;
+
+
+public class MailboxAssignmentVisitor implements StageNodeVisitor<Void, DispatchablePlanContext> {
+  public static final MailboxAssignmentVisitor INSTANCE = new MailboxAssignmentVisitor();
+
+  @Override
+  public Void visitAggregate(AggregateNode node, DispatchablePlanContext context) {
+    node.getInputs().stream().forEach(input -> input.visit(this, context));
+    return null;
```
Comment (walterddr) [inline pinot-query-planner/src/main/java/org/apache/pinot/query/planner/physical/MailboxAssignmentVisitor.java:46]: we can simply have a private visit(StageNode node) method to catch all others. here only MailboxSend and Receive is required to visit.

File: pinot-common/src/main/proto/worker.proto [hunk 3]
```diff
@@ -72,15 +72,28 @@ message StagePlan {
...
   string virtualAddress = 2;
   StageNode stageRoot = 3;
-  repeated StageMetadata stageMetadata = 4;
+  StageMetadata stageMetadata = 4;
 }
 
 message StageMetadata {
   repeated WorkerMetadata workerMetadata = 1;
   map<string, string> customProperty = 2;
+  map<int32, StageMailboxInfo> stageMailboxInfoMap = 3;
```
Comment (walterddr) [inline pinot-common/src/main/proto/worker.proto:81]: let's put this in the worker metadata.

File: pinot-common/src/main/proto/worker.proto [hunk 4]
```diff
@@ -72,15 +72,28 @@ message StagePlan {
...
+  string mailboxId = 1;
+  string mailboxType = 2;
+  string mailboxHost = 3;
+  int32 mailboxPort = 4;
+  map<string, string> customProperty = 5;
+}
+
+message StageMailboxInfo {
+  repeated MailboxInfo mailboxInfo = 1;
+}
```
Comment (walterddr) [inline pinot-common/src/main/proto/worker.proto:99]: this makes hard to convert mailbox info going forward. I would suggest directly modifying the WorkerMetadata
```
message WorkerMetadata {
  string virtualAddress = 1;
  map<int32, MailboxMetadata> mailboxMetadata = 2;
  map<string, string> customProperty = 3;
}

message MailboxMetadata {
  repeated string mailboxId = 1;
  repeated string virtualAddress = 2;
  map<string, string> customProperty = 3;
}
```

Comment (xiangfu0) [reply pinot-common/src/main/proto/worker.proto]: Done

File: pinot-query-planner/src/main/java/org/apache/pinot/query/planner/physical/MailboxAssignmentVisitor.java (hunk 2 above)
Comment (xiangfu0) [reply pinot-query-planner/src/main/java/org/apache/pinot/query/planner/physical/MailboxAssignmentVisitor.java]: done

File: pinot-query-planner/src/main/java/org/apache/pinot/query/routing/StageMetadata.java (hunk 1 above)
Comment (xiangfu0) [reply pinot-query-planner/src/main/java/org/apache/pinot/query/routing/StageMetadata.java]: yes, changed to list, the ordering has to be kept.

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/plan/serde/QueryPlanSerDeUtils.java [hunk 5]
```diff
@@ -98,16 +93,33 @@ private static StageMetadata fromProtoStageMetadata(Worker.StageMetadata protoSt
...
+    builder.putAllMailBoxInfosMap(protoToMailbox(protoWorkerMetadata.getMailboxMetadataMap()));
     builder.putAllCustomProperties(protoWorkerMetadata.getCustomPropertyMap());
     return builder.build();
   }
 
-  public static List<Worker.StageMetadata> stageMetadataListToProtoList(List<StageMetadata> stageMetadataList) {
-    List<Worker.StageMetadata> protoList = new ArrayList<>();
-    for (StageMetadata stageMetadata : stageMetadataList) {
-      protoList.add(toProtoStageMetadata(stageMetadata));
+  private static Map<Integer, List<MailboxMetadata>> protoToMailbox(
```
Comment (walterddr) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/plan/serde/QueryPlanSerDeUtils.java:101]:
```suggestion
  private static Map<Integer, List<MailboxMetadata>> fromProtoMailboxMap(
```

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/plan/serde/QueryPlanSerDeUtils.java [hunk 6]
```diff
@@ -122,7 +134,37 @@ private static Worker.StageMetadata toProtoStageMetadata(StageMetadata stageMeta
...
+    return builder.build();
+  }
+
+  private static String toMailboxCustomPropertiesKeyPrefix(String mailBoxId, String key) {
+    return String.format("%s.%s", mailBoxId, key);
+  }
+
+  private static String fromMailboxCustomPropertiesKeyPrefix(String mailBoxId, String key) {
+    return key.split(mailBoxId + "\\.")[1];
+  }
```
Comment (walterddr) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/plan/serde/QueryPlanSerDeUtils.java:169]: for these mailbox specific customProperty utils we can put them in MailboxMetadata class IMO (to make it consistent with Stage and Worker level metadata class)

File: pinot-query-planner/src/main/java/org/apache/pinot/query/planner/physical/DispatchablePlanVisitor.java [hunk 7]
```diff
@@ -60,10 +60,17 @@ public QueryPlan constructDispatchablePlan(StageNode globalReceiverNode,
     dispatchablePlanContext.getDispatchablePlanStageRootMap().put(0, globalReceiverNode);
     // 3. add worker assignment after the dispatchable plan context is fulfilled after the visit.
     computeWorkerAssignment(globalReceiverNode, dispatchablePlanContext);
-    // 4. convert it into query plan.
+    // 4. compute the mailbox assignment for each stage.
+    computeMailboxAssignment(dispatchablePlanContext);
+    // 5. convert it into query plan.
```
Comment (walterddr) [inline pinot-query-planner/src/main/java/org/apache/pinot/query/planner/physical/DispatchablePlanVisitor.java:65]: Let's add a TODO here for the 2 methods. later we want to refactor these into pluggable algorithms. (e.g. worker assignment visitor and mailbox assignment visitor)

but so far the entrypoint looks super clean and i love it

**Review by walterddr — APPROVED** [review]
lgtm otherwise. please kindly follow up with the comments.

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/plan/serde/QueryPlanSerDeUtils.java (hunk 6 above)
Comment (xiangfu0) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/plan/serde/QueryPlanSerDeUtils.java:169]: I feel there will be per mailbox metadata. Or shall I make the mailboxMetadata from: 
` map<int32, MailboxMetadata> mailboxMetadata = 2;`
to something like ` map<int32, repeated MailboxMetadataList> mailboxMetadataMap = 2;` ?

File: pinot-query-planner/src/main/java/org/apache/pinot/query/planner/physical/DispatchablePlanVisitor.java (hunk 7 above)
Comment (xiangfu0) [reply pinot-query-planner/src/main/java/org/apache/pinot/query/planner/physical/DispatchablePlanVisitor.java]: added comments

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/plan/serde/QueryPlanSerDeUtils.java (hunk 6 above)
Comment (xiangfu0) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/plan/serde/QueryPlanSerDeUtils.java]: Made proto and java class in sync for mailboxMetadata.

**Pull request merged**